types/wlr_surface_layers.c

```diff
@@ -0,0 +1,287 @@
+			"Unknown input device in input_types");
+		return;
+	}
+	if (exclusive_types >> WLR_LAYER_SURFACE_INPUT_DEVICE_COUNT) {
+		wl_resource_post_error(resource,
+			LAYER_SURFACE_ERROR_INVALID_INPUT_DEVICE,
+			"Unknown input device in exclusive_types");
+		return;
+	}
+
+	surface->pending->input_types = input_types;
+	surface->pending->exclusive_types = exclusive_types;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_INTERACTIVITY;
+}
+
+static void layer_surface_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+
+	if (anchor >> 4) {
+		wl_resource_post_error(resource, LAYER_SURFACE_ERROR_INVALID_ANCHOR,
+			"Unknown anchor");
+		return;
+	}
+
+	surface->pending->anchor = anchor;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_ANCHOR;
+}
+
+static void layer_surface_set_exclusive_zone(struct wl_client *client,
+		struct wl_resource *resource, uint32_t zone) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->pending->exclusive_zone = zone;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_EXCLUSIVE_ZONE;
+}
+
+static void layer_surface_set_margin(struct wl_client *client,
+		struct wl_resource *resource, int32_t horizontal, int32_t vertical) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->pending->margin_horizontal = horizontal;
+	surface->pending->margin_vertical = vertical;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_MARGIN;
+}
+
+void wlr_layer_surface_move_state(struct wlr_layer_surface_state *next,
+		struct wlr_layer_surface_state *state) {
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_INTERACTIVITY) {
+		state->input_types = next->input_types;
+		state->exclusive_types = next->exclusive_types;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_ANCHOR) {
+		state->anchor = next->anchor;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_EXCLUSIVE_ZONE) {
+		state->exclusive_zone = next->exclusive_zone;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_MARGIN) {
+		state->margin_horizontal = next->margin_horizontal;
+		state->margin_vertical = next->margin_vertical;
+	}
+
+	state->invalid |= next->invalid;
+	next->invalid = 0;
+}
+
+void wlr_layer_surface_get_position(struct wlr_layer_surface *layer_surface,
+		double *x, double *y) {
+	int width = layer_surface->surface->current->width;
+	int height = layer_surface->surface->current->height;
+	struct wlr_layer_surface_state *state = layer_surface->current;
+
+	int output_width, output_height;
+	wlr_output_effective_resolution(layer_surface->output, &output_width,
+		&output_height);
+
+	if (state->anchor & LAYER_SURFACE_ANCHOR_LEFT) {
+		*x = state->margin_horizontal;
+	} else if (state->anchor & LAYER_SURFACE_ANCHOR_RIGHT) {
+		*x = output_width - width - state->margin_horizontal;
+	} else {
+		*x = (double)(output_width - width) / 2;
```

Comment:
According to the spec, you should consider the margin here when placing in the center.

Reply:
>Requests that the surface be placed some distance away from the anchor point on the output, in pixels.

If there's no anchor point for an axis, I don't think it makes sense to take into account a margin.

Reply:
Don't worry about it in the center. We can update the spec if it's confusing.